lib/puppet/type/package.rb

```diff
@@ -296,6 +298,61 @@ def self.title_patterns
       [ [ /(.*)/m, [ [:name] ] ] ]
     end
 
+    # I am accepting the `environment` parameter using the same format
+    # that the `exec` type does, but I transform it to a hash in the
+    # parameter's `munge` code.
+    #
+    # I am skipping the :required_features in the declaration, since
+    # that causes the parameter to be [mostly] silently ignored.  I will
+    # enforce the existence of the :settable_environment feature during
+    # validation, instead.
+    #   newparam(:environment, :required_features=>:settable_environment) do
+    newparam(:environment) do
+      desc "Any additional environment variables you want to set for operations
+        against this package.  The value should be a single string of the form
+        `NAME=value` or an array of such strings.
+
+        An error will be raised if this parameter is used with a provider that
+        does not support it."
+
+      validate do |values|
+        # enforce provider feature here
+        unless provider.satisfies?([:settable_environment])
+          raise ArgumentError, "environment is not supported by the #{provider.class.name} provider"
+        end
+        values = [values] unless values.is_a? Array
```

Comment:
I tend to prefer the `values = [values].flatten` format, so that we're not tied to objects being specific ruby types in the future.